[Rezwan Ahmed] iP

RezwanAhmed123 opened the pull request:
## Mortal Reminder
> A sobering reminder that everyone is a mortal and healing doesn't mean immortality (League of Legends).

Mortal Reminder gives you so much power to remember anything. Forget having a brain, its unreliable.

Reasons to ~~not using your brain~~ MortalReminder:
* **BEAUTIFUL** graphical user interface (GUI)
* easy to learn and use
* _QUICK_ loading and unloading

How to use?
1. download it from [here](https://nus-cs2103-ay2425s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

Features:
- [ ] managing things to do
- [ ] managing events
- [ ] managing deadlines
- [ ] finding saved events
- [ ] checking upcoming events 👀
---
Here's a snippet of the launcher class that you can look at and practice:

``` java
/**
 * A launcher class to workaround classpath issues.
 */
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

RezwanAhmed123 committed:
Completed Level-0. Renamed the chatbot to Mortal Reminder.

RezwanAhmed123 committed:
Updated the README to remove template information and replace it with my own placeholders.

RezwanAhmed123 committed:
Updated Level-0. Changed all non-main methods from static to non-static types.

RezwanAhmed123 committed:
Updated MortalReminder.java with completed Level-1.

RezwanAhmed123 committed:
Added List.java.

RezwanAhmed123 committed:
Updated ToDoList.java.

RezwanAhmed123 committed:
Updated MortalReminder.java and completed Level-2.

RezwanAhmed123 committed:
Added Task.Java

RezwanAhmed123 committed:
Renamed ToDoList.java to Storage.java in occurrences and files.

RezwanAhmed123 committed:
Updated Task.java

RezwanAhmed123 committed:
Added a new line at the end of Task.java file.

RezwanAhmed123 committed:
Updated Storage.java

The list inside storage only stores Strings which are primitive structures.

Primitive Structures does not allow for complex representation of data which is what we eventually want to achieve with the Todos, Events and Deadlines in this project.

Thus, the list structure in the storage class needs to be modified and the accompanying methods also need to be fixed accordingly to allow a more complex data structure,  the Task class, to be stored in the list.

This allows each item inside the list to be packaged with both the task name and the mark status together in one object, and eventually we can extend this to other information we want to store together instead of separately.

RezwanAhmed123 committed:
Renamed Storage.java to TaskList.java

The name storage should be reserved for dealing with loading and saving tasks in the file.

The current structure inside the file TaskList.java contains the task list itself and has operations to add/delete tasks in the list.

I will eventually separate the two but since this file already contains the tasklist itself, i will proceed to keep rename it as TaskList.java.

This will prevent further confusion in future and is more intuitive for the reader of this code to follow.

RezwanAhmed123 committed:
Enabled small scale regression testing.

Regression testing manually is time consuming and laborious. This problem will only become more difficult to solve as the codebase increases in size.

I automated test scripts, expected input, expected output txt files were modified with relevant regression testing cases. I also updated the .sh file with the new name of the main file to allow the testing.

This will automate the testing process and save time.

RezwanAhmed123 committed:
Added AI.md

I forsee the use of AI to autocomplete quite alot of methods which are similar to each other.

Many of the methods which I forsee creating are kind of repetitive and time consuming in the next few levels.

As such I have created AI.md to explain my rationale for using AI tools in a widespread manner and the explanations behind each use.

This will allow me to document my AI tools used concurrently with the progress of the project, to ensure adherence to the rules of the use of AI in the project.

RezwanAhmed123 committed:
Added Parser.java

There is many different parts of the code dealing with making sense of the user input.

This makes the debugging process and the code itself very messy.

In order to modulate the code better, I have created Parser.java to only handle all of the user input and change it to a state that the program can understand.

This will allow centralised handling of user inputs and better management of commands by the program.

RezwanAhmed123 committed:
Added Command.java and CommandExecutor.java

I noticed the commands starting to become more and more complex.

More complex commands need to be represented by their own classes
to ensure the data is well modulated and communicated within the program
more efficiently.

Lets create a class to package command information together for easier storage
and communication and another class CommandExecutor to help execute such
commands instead of doing everything within the Parser class.

This ensures a cleaner code base, with all information referring to
commands and their executions within these files while the parser only handles
parsing user input. This will help with easier debugging due to better modularity.

RezwanAhmed123 committed:
Added Command.java and CommandExecutor.java

I noticed the commands starting to become more and more complex.

More complex commands need to be represented by their own classes
to ensure the data is well modulated and communicated within the program
more efficiently.

Lets create a class to package command information together for easier storage
and communication and another class CommandExecutor to help execute such
commands instead of doing everything within the Parser class.

This ensures a cleaner code base, with all information referring to
commands and their executions within these files while the parser only handles
parsing user input. This will help with easier debugging due to better modularity.

RezwanAhmed123 committed:
Merge branch 'master' of https://github.com/RezwanAhmed123/ip

# Conflicts:
#	src/main/java/Command.java
#	src/main/java/CommandExecutor.java

RezwanAhmed123 committed:
Added FormattedPrinting.java

There is too many printing instances in many different files.

The formatting of the output is not always consistent and the handling of the printing is very messy for the coder.

Lets create a new printing class to handle all of the outputs from the chatbot. Ensure that the class is static so
that new instances will not have to be called and the printing can be done by invoking the class name only.

Using a static class to handle all the printing ensures that there is only one operation of printing
and multiple instances of printing does not occur within the program, leading to messy outputs.
It also makes it easier to use the printer anywhere without having to instantiate it leading to
cleaner code and debugging.

RezwanAhmed123 committed:
Updated all files with new Format Print.

All these files had some printing involved and were inconsistent in formatting.

Lets modify them to ensure consistent output generated by the chatbot by incorporating
the FormattedPrinting class.

RezwanAhmed123 committed:
Updated gitignore to ignore .run files.

The .run files are different for some of my devices and can be generated on each device anyways.

Since the file can be generated on each device anyways, there is no need to track it on git and can
be put inside the .gitignore file

RezwanAhmed123 committed:
Add Deadline.java, Events.java, ToDo.java

The Task class has the common behaviour for all the different types
of items that can be found inside of the TaskList inside the chatbot.

However, it cannot capture all possible types of behaviours of the
many different types of tasks such as ToDos, Deadlines and Events
which all need to be configured differently.

Let's use polymorphism to extend the task class to these 3 different
classes. Let's change Task class to an abstract class with an abstract
method "getDesctiption" which is to be implemented differently by
each different type of Task. The constructor will also be shortened
to allow each type of Task to store the description differently.
Lastly, lets change the CommandExecutor file to call these different
types of tasks under each different types of commands.

Using polymorphism allows for the 3 different classes to have common
behaviours that do not need to be recoded at any minor change while
also be called using a common object type (Task) especially when we
are not sure what type of object will be returned. This also gives
us flexibility to change the behaviour of each class independently
from one another.

RezwanAhmed123 committed:
Minor Commit: modified ToDo.java

Changed the Constructor to include the description during
initialisation.

RezwanAhmed123 committed:
Add CommandTypes.java

Long switch statement inside of Parser.java is not ideal.

It needs to be manually modified every time a new comment is added.
This means we are repeating the change every time we were to add a
new command both to the Enum List inside the CommandExecutor file
and the Parser file.

Lets create a separate Enum file to abstract out all the possible
commands and create a method inside Parser.java to search for the
command inside this Enum class.

Using enum's is preferable than a bunch of switch statements as it
reduces the amount of recoding necessary when new commands are
implemented and it also makes the code look cleaner to readers.

RezwanAhmed123 committed:
Add Error Handling

The program is not built to handle errors correctly.

This causes the program to crash if there is an input error by the
user, thus the program is not able to provide feedback to the user
about their input.

Let's add the following error handling functionality to the chatbot:
- marking/unmarking error: when the user tries to mark something
  that has already been marked/unmarked
- trying to view or mark items in an empty list
- adding a task without anything
- adding incorrect number of inputs for todo, deadline and events
  respectively.
- entering non-integers after mark/unmark
- marking/unmarking items out of the bounds of the list.

With the following error handling, the chatbot becomes more robust
and is able to provide useful feedback to the user about their input
so that the user is able to self correct their mistakes.

RezwanAhmed123 committed:
Add delete task functionality

There is no way to delete tasks after you are done with them.

This makes the list look very messy after a few tasks have been
completed.

Let's create a delete functionality to allow users to delete tasks.

This will make the list easier to organise for the user, especially
if they have completed a task or added the task wrongly.

RezwanAhmed123 committed:
Add Semi-Automated Text UI Testing

The amount of things to test per new function added is increasing.

This makes regression testing very laborious and time consuming.

Lets set-up semi automated testing using expected and input text
files to speed up the testing process.

This will make future testing quicker and less labour intensive.

RezwanAhmed123 committed:
Reformat Code

Reformatted code to follow the code style required for CS2103 module
guidelines.

RezwanAhmed123 committed:
Allowed Long-term Storage

The list created inside of the CLI is only able to keep track of one
session and the list is destroyed after the session ends.

This makes the app useless unless the user keeps it open all the
time and if the app is deleted, the user loses all access to what
he/she has done.

Lets create a Storage class to enable long term storage and give the
user the power to decide when to delete tasks even after closing the
session.

Long term storage makes the app more robust in case of accidental
closing of the CLI app and eventually the real app.

RezwanAhmed123 committed:
Add load and modify delete functionality

The implementation of the long-term storage is incomplete due to
inability to load and delete properly from the long term storage
file.

This makes the app unable to access the information it has stored
and essentially be the same as before the long term storage.

Lets add load and delete from file functionality. Lets also add a
clear functionality to aid with the delete functionality.

The long term storage will now be completely accessible by the app
and be able to support memory between sessions.

RezwanAhmed123 committed:
Add Check if Storage Directory exists

The program only checked if the file existed not the directory.

If the directory did not exist, the program would have thrown an
error on the first opening of the app itself.

Lets add a check to see if the directory exists when loading the app
at the start of every session.

This makes the system more robust on initial opening and even if the
directory was deleted by accident.

RezwanAhmed123 committed:
Merge branch 'branch-Level-7'

RezwanAhmed123 committed:
Fix runtest files

Minor fix to ensure runtest files working as intended after
implementing the check and create directory on top of file.

RezwanAhmed123 committed:
Add time readability

Deadlines and Event times are stored as string.

Program is unable to verify or use the time inputs from user in any
meaningful way.

Lets add time functionality to these two classes by using the
LocalDateTime package from java.

By storing times as LocalDateTime objects, these objects can be used
for more meaningful interactions with the user in future commands
such as looking up upcoming deadlines or events. This also serves as
a check for valid dates and times for events.

RezwanAhmed123 committed:
Completed Level-8

Time was merely stored and output in terms of LocalDateTime objects.

There is potential for more uses of said objects as per last commit.

Lets implement a more meaningful use of said data by printing all
upcoming deadlines and events only if they are unmarked.

With this, users will be able to check all upcoming tasks which they
have not done in the event that the list is very huge.

RezwanAhmed123 committed:
Merge branch 'branch-Level-8'

RezwanAhmed123 committed:
Minor code formatting commit

RezwanAhmed123 committed:
Separate into packages

The files are not divided into proper packages.

This makes it difficult for any reader to understand which types are
related, where to find which type of classes. It also can cause
namespace conflicts if the number of classes were to increase and
does not encapsulate data well within certain types of files.

Lets separate the files into packages, with the main package being
mortal_reminder, and inside that package we will have 4 subpackages,
namely backend, commands, io and tasks. The main method will remain
inside of the main package as a standalone file.

With these changes, the backend, the tasks, command types and the
user interaction files will all be separated within their own
package folders and the programmer can easily find and change files
they need to in an intuitive manner.

RezwanAhmed123 committed:
Merge branch 'branch-A-packages'

RezwanAhmed123 committed:
Merge branch 'add-gradle-support'

RezwanAhmed123 committed:
Add gradle support

The app needs to be built and tested manually right now.

Testing and building are two of the things which can be automated
in a project to lighten the load of a programmer as well as ensure
build consistency.

Lets add gradle support for the project.

With gradle support, right now, we can build the project automatically
and run it automatically.

RezwanAhmed123 committed:
Merge branch 'branch-A-Gradle'

RezwanAhmed123 committed:
Add JUnit Test for TaskList getTask method

We have been testing the entire application in entirety so far, and
not specific methods to ensure it is not buggy.

This makes it difficult to pinpoint sources of error in the code and
increases the overall debugging time.

Lets create JUnit tests to test the getTask method inside of
TaskList.java. This method is integral to the application because
it is a key part of data storing and manipulation. If this method is
buggy, it may be easy to notice in the results but time-consuming to
check.

Thus, lets create an automated method to test this method to save
debugging time in future.

RezwanAhmed123 committed:
Add Storage.java test for loadTaskListFromFile

We have been testing the entire application in entirety so far, and
not specific methods to ensure it is not buggy.

This makes it difficult to pinpoint sources of error in the code and
increases the overall debugging time.

Lets create JUnit tests to test the loadTaskListFromFile method
inside of Storage.java.

This method is integral to the application because it is ensures the
long-term memory of the app. If this method is buggy, it may not be
very obvious at first glance, unless tested further in depth.

Thus, an automated method to test this method was created to save
debugging time in future.

RezwanAhmed123 committed:
Merge branch 'branch-A-JUnit'

RezwanAhmed123 committed:
Create JAR file

Completed gradle automation of building a JAR file with each build.

RezwanAhmed123 committed:
Add Javadocs

All methods were not explained adequately inside the code.

This can lead to future programmers being unable to understand how
to modify the existing code making the code difficult to be improved
on.

Lets add javadocs to as many methods and classed as possible
explaining their functionality succinctly.

This will allow other programmers reading the code to understand and
work with the code or even create tests for the code more easily.

RezwanAhmed123 committed:
Formatted code to follow a consistent code style

There is some inconsistency in the import lines and the code style
for some of the files.

This makes the code slightly unreadable to the user.

Lets reformat the code using the IDE automation and follow the
recommended coding style in the CS2103 coding style recommendations.

This will make the code alot more readable to the user.

RezwanAhmed123 committed:
Add find functionality

The list of tasks an actual user will work with is usually rather
large.

This makes it sometimes difficult to find specific task that you have
added in before.

Lets add search functionality for users to be able to use to find
all tasks with similar description to what they have searched for.

This will make the app more easy for users to use in a real
application.

RezwanAhmed123 committed:
Merge branch 'branch-A-JavaDoc'

RezwanAhmed123 committed:
Merge branch 'branch-A-CodingStandard'

RezwanAhmed123 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/mortal_reminder/backend/TaskList.java

RezwanAhmed123 committed:
Modified mark and unmark

Marking and Unmarking tasks are not captured in the storage.

If a user were to mark the task and close the app, the change would
not be captured inside of the storage file and would not be reflected
in the app the next time it is turned on.

Lets add the refreshStorage method to the mark and unmark methods to
ensure the changes are kept each time the app is closed.

That should fix any long term memory loss issues with the mark and
unmark methods.

RezwanAhmed123 committed:
Format changes

Minor format changes before starting on new weeks work

RezwanAhmed123 committed:
Integrate checkstyle

Add checkstyle to the project to track objective style issues to
automatically be ironed out.

Some of the current java coding standard violations may not be
easily caught by the developer.

These coding violations may go unnoticed and cause inconsistency
down the line.

Let's integrate CheckStyle to automatically check and help the
developer to rectify any coding violations real time, saving a lot
of time down the line.

RezwanAhmed123 committed:
Merge branch 'branch-A-CheckStyle'

RezwanAhmed123 committed:
Add GUI

Command Line App is not appealing to many people who may not use the
app just because the user interface (command line is difficult to use)

The command line also makes it difficult to see history of the chat
with the bot because it does not provide a good text field for the user
to use.

Lets add a GUI using JavaFX to our app to allow better interaction
with the user. Build just the basic app (interaction with the user
only, no welcome message) first.

This will make the app more appealing to potential users and make
interaction much easier.

RezwanAhmed123 committed:
Complete GUI

The GUI did not have a welcome message, so it seemed very abrupt
when the application started.

This may seem confusing to the reader when they use it for the first
time and be unable to understand what to do next.

Let's add some formatting and a welcome message to the app.

This allows the user to have some form of understanding of what the
app is about to do and have a more intuitive start to the application.

RezwanAhmed123 committed:
Merge branch 'branch-Level-10'

RezwanAhmed123 committed:
Modify JavaFX version to match runtime version

There is a mismatch between the javafx versions in the .xml files
and the build.gradle file.

This causes error messages to show up and may cause issues in
code behaviour down the road.

Lets modify both javafx versions to be '17.0.7'.

This will ensure consistency in javafx versions being used.

RezwanAhmed123 committed:
Add Data folder to Resources

List Storage file is left as a standalone file inside of the
resources folder.

This is messy for code organisation and readability and reduces
any improvement potential by other developers in future.

Lets create a folder 'data' to keep the storage list file for better
organisation.

This will make the code more readable and intuitive for future
developers.

RezwanAhmed123 committed:
Format code according to code style

There is some error in the order of imports and missing spaces.

This makes the code unreadable to future developers.

Lets reformat the order of imports to make it look more readable.

This will make the code base more consistent and usable for future
developers.

RezwanAhmed123 committed:
Add Varargs to find functionality

Find is restricted to only one search phrase.

This makes it inflexible to search for things that may be related
but not with the same phrases. For example, read book and write
report are both within similar domains of study but do not share
similar phrases. As such, there needs to be a way to quick search
them.

Lets add Varargs to the find function, and allow different search
terms to be separated with a comma.

This allows similar types of things to be searched together.

RezwanAhmed123 committed:
Merge branch 'branch-A-Varargs'

RezwanAhmed123 committed:
Add dynamic background colour for DialogBox

The background colours for the DialogBox are pretty uninteresting as
they are all green with black text.

The user is unable to immediately tell if the command they sent was
indeed recorded correctly or not as there is no visual feedback.

Lets add dynamic colouring to the dialogbox so that there is visual
feedback on the type of command that was just processed by the user.

This will give the user a better user experience and ability to
instantly recognise if they had made an error in passing a command
previously just by the colour of the dialogbox.

RezwanAhmed123 committed:
Modify JavaDocs

Some methods' Javadocs were not updated after the GUI was added.

These methods changed from being void (due to the use of System.out.
println) but changed to returning a string which then is printed on
the GUI.

This makes the Javadoc outdated and difficult for future developers
to refer to.

Lets update the Javadocs to make the code more readable and
understandable.

This would then allow future developers to understand the use of the
modified methods with no confusion.

yanqiyqh commented on AI.md:
```diff
@@ -0,0 +1,52 @@
+### What worked and what did not
+Making the prompt specific to what I am doing, with some examples of my task, tasklist and parser files helped the 
+generated code be very close to what I am looking to create and it saved a lot of time too. I think my prompts were too 
+specific in some cases so the AI started hard coding values in, which I corrected myself.
```
nice

yanqiyqh reviewed (commented):
LGTM

ethantyh692 commented on src/main/java/mortalreminder/MortalReminder.java:
```diff
@@ -0,0 +1,35 @@
+        return FormattedPrinting.welcome();
+    }
+
+    public Processor getProcessor() {
+        return this.processor;
+    }
+
+    public TaskList getTaskList() {
+        return this.taskList;
+    }
```
I like the use of verbs for getters

ethantyh692 commented on src/main/java/mortalreminder/backend/Processor.java:
```diff
@@ -0,0 +1,171 @@
+                .reduce((accumulator, element) -> accumulator + " " + element)
+                .orElse("");
+
+        switch (commandType) {
+        case LIST:
+            return FormattedPrinting.printList(taskList);
+
+        case FIND:
+            TaskList foundTasks = taskList.findTasks(commandDetails.split(","));
+            return FormattedPrinting.printSimilarTasks(foundTasks);
+
+        case MARK:
+        case UNMARK:
+        case DELETE:
+            return markUnmarkOrDelete(commandDetails, taskList, commandType);
+
+        case TODO:
+        case DEADLINE:
+        case EVENT:
+            return createTask(commandDetails, taskList, commandType);
+
+        case CLEAR_TASKS:
+            Storage.clearListFile();
+            return taskList.clearList();
+
+        case UPCOMING_TASKS:
+            return upcomingTasks(taskList);
+
+        default:
+            return FormattedPrinting.unknownCommand();
+        }
```
I like how this switch, case block adheres to the coding standards

ethantyh692 commented on src/main/java/mortalreminder/backend/TaskList.java:
```diff
@@ -0,0 +1,155 @@
+        return this.taskList.get(index);
+    }
+
+    public boolean isEmpty() {
+        return this.taskList.isEmpty();
+    }
```
I like how this boolean method sounds like a boolean

ethantyh692 commented on src/main/java/mortalreminder/io/FormattedPrinting.java:
```diff
@@ -0,0 +1,228 @@
+    /**
+     * Returns the welcome message to the user upon program startup.
+     */
+    public static String welcome() {
+        String welcomeMessage = "Hello I'm Mortal Reminder!\n"
+                + "What can I do for you?";
+        return getResponse(welcomeMessage);
+    }
```
I like the line break as it improves readability.

ethantyh692 commented on README.md:
```diff
@@ -1,24 +1,3 @@
-# Duke project template
+# Mortal Reminder Chatbot Project!
```
Overall the code is very clean, organised and adheres with the coding standards. Good job!

ethantyh692 reviewed (commented):
LGTM

RezwanAhmed123 committed:
Add MortalReminderException class

The exception handling is not consistent throughout the entire file.

This prevents specific debugging methods to each specific problems
and also makes the code more difficult to follow.

Let's add a new Exception class to handle all the exceptions in the
current iteration of the project.

This makes the code more readable and future exceptions easier
to add.

RezwanAhmed123 committed:
Continue reformating exceptions

Not all exceptions have been reformatted with the new class.

This makes the previous commit incomplete.

Lets complete reformatting all the exceptions into the new exception class.

This will complete the exception handling thus far.

ethan-goh commented on src/main/java/mortalreminder/io/FormattedPrinting.java:
```diff
@@ -0,0 +1,161 @@
+                + task.getDescription();
+    }
+
+    /**
+     * Returns a confirmation message when the user adds a new task.
+     *
+     * @param task     {@link Task} that has been added to the {@link TaskList}
+     * @param taskList The current list of all tasks tracked by the backend.
+     * @return string of confirmation message.
+     */
+    public static String addTask(Task task, TaskList taskList) {
+        String message = "Got it. I've added this task:\n"
+                + printTask(task)
+                + "\n"
+                + "Now you have "
+                + taskList.getSize()
+                + " task(s) in the list.";
+        return getResponse(message);
+    }
+
+    /**
+     * Returns a confirmation message when the user deletes a task.
+     *
+     * @param task     {@link Task} that has been deleted from the {@link TaskList}
+     * @param taskList taskList The current list of all tasks tracked by the backend.
+     * @return string confirmation message of a successful deletion of task.
+     */
```
I like the use of JavaDocs in this class

ethan-goh commented on src/main/java/mortalreminder/tasks/Task.java:
```diff
@@ -0,0 +1,114 @@
+public abstract class Task {
+    protected String description;
+    protected String type;
+    protected boolean isDone;
```
Good naming of boolean value

ethan-goh commented on src/main/java/mortalreminder/tasks/Deadline.java:
```diff
@@ -0,0 +1,122 @@
+        this.description = descriptionString[0].trim();
+        this.deadline = getTime(descriptionString[1].trim());
+        this.type = "D";
+
```
Possible to delete this whitespace as it is unnecessary

ethan-goh commented on src/main/java/mortalreminder/io/MortalReminderWindow.java:
```diff
@@ -0,0 +1,69 @@
+package mortalreminder.io;
+
+import javafx.fxml.FXML;
+import javafx.scene.control.Button;
+import javafx.scene.control.ScrollPane;
+import javafx.scene.control.TextField;
+import javafx.scene.image.Image;
+import javafx.scene.layout.AnchorPane;
+import javafx.scene.layout.VBox;
+import mortalreminder.MortalReminder;
+import mortalreminder.commands.Command;
+import mortalreminder.commands.CommandTypes;
```
Import statements are ordered consistently, a nit would be to separate the mortal reminder and javafx imports.

ethan-goh reviewed (commented):
LGTM

RezwanAhmed123 committed:
Reformat code

The code is not completely adherent to the code style and
conventions inside of the CS2103 requirements.

This makes the code seem like it is not consistently done by one
person.

Lets modify some parts of the code not adherent to the codestyle.

This will make the code more consistent and readable.

RezwanAhmed123 committed:
Format code

The current code does not completely adhere to the requirements of the course.

Lets format some of the code so that the coding standard violations are fixed and
the code looks more consistent.

This will make the code more readable and consistent to the coding standard of
the course.

RezwanAhmed123 committed:
Clean up Code

There are some violations of the naming convention and javadocs.

This makes the code abit confusing to follow.

Let's reformat and cleanup both the javadocs and the method names to
make the code alot more readable for future users.

RezwanAhmed123 committed:
Add Assertion to TaskList.java

There is a method deleteTask() which holds an assumption that has
has not been explicitly shown in the code. Delete task will always
only be run once there are tasks that have been added to the
taskList in the first place and have already been checked for
validity in the addTask or loadTask method call.

By not showing this assumption to the future developers, they may
not be aware of this and try to change the method in an unintended
way, which may break the delete function.

Let's create an assertion at the else statement inside the
deleteTask method that ensures the code does not ever reach that
location.

This will help new developers understand the code better AND it
becomes more apparent when this assumption is no longer true,
prompting the developer to modify other methods like the addTask
and loadTask method such that the assumption becomes true again.

RezwanAhmed123 committed:
Add assertions to Processor.java

There is a fair bit of assumptions made inside the methods of
Processor.java, specifically in the markUnmarkOrDelete and
createTask methods.

In both methods, if-else statements were used to call on the last
conditional in the method but if the handleCommand method was to be
modified, there is no guarantee that this assumption will hold.

Let's add these assumptions to ensure that the developer knows about
the said assumptions and modifies other methods in the project
around these assumptions.

This will make the project less confusing to work on in the future.

RezwanAhmed123 committed:
Merge pull request #1 from RezwanAhmed123/branch-A-Assertions

Add assertions to Processor.java and TaskList.Java

RezwanAhmed123 committed:
Modify Code Quality

Some code blocks are long and nested and some magic numbers were
used.

This affects readability of code.

Lets reduce the nested code inside of the Processor.java,
Storage.java and Parser.java. Lets add some comments
to explain the magic letters inside of Todo.java. Deadline.java
and Event.java.

This makes the code alot more understandable to future developers.

RezwanAhmed123 committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/mortalreminder/backend/Processor.java

RezwanAhmed123 committed:
Improve code quality

The current code quality is decent but there is room for improvement.

There are some complex methods and there is room to make the happy
path more apparent in the code.

Lets reduce the nesting in these complicated methods specifically
in the Processor.java to make the happy path more apparent.

This makes the code alot more readable and less confusing for future
developers.

RezwanAhmed123 committed:
Code Refactoring

There is a lot of accumulated technical debt as until now, the focus
was on getting the code to work properly.

This made some of the code be redundant and other parts of the code
to be very long.

Let's refactor the code and re-organise the methods and classes to
ensure that the code organisation and quality improves.

This will reduce the technical debt accumulated, making the codebase
more useable for the future.

RezwanAhmed123 committed:
Improve code quality

RezwanAhmed123 committed:
Merge pull request #2 from RezwanAhmed123/branch-A-CodeQuality

Code Refactoring

RezwanAhmed123 committed:
Merge branch 'master' into branch-A-Streams

RezwanAhmed123 committed:
Add Streams to find task function

The function has two nested for-loops.

This may affect code quality down the line.

Let's implement the method using streams to improve the readability
and clarity of the method to other developers.

This will make the code look a lot more clean.

RezwanAhmed123 committed:
Merge branch 'branch-A-Streams'

RezwanAhmed123 committed:
Add .github/workflows/gradle.yml

This allows for continuous integration of the project inside of
github.

RezwanAhmed123 committed:
Merge branch 'branch-A-CI'

RezwanAhmed123 committed:
Add newline to end of gradle.yml

RezwanAhmed123 committed:
Add flexible inputs

The commands are very rigid and not customisable.

This increases user frustration when they misspell a single word in
a long command.

Let's add functionality to make the commands more flexible and be
called on using different words.

This makes the user experience better and more customised to each
user.

RezwanAhmed123 committed:
Merge branch 'branch-BCD-Extension'

RezwanAhmed123 committed:
Improved First Startup Experience

The necessary storage files were not being created on first start of
the app.

The user will be confused at the error message and not want to use
the app anymore.

Let's improve the first startup experience and enable all necessary
file creations upon the first creation of the app.

This will make the initial user experience much better and smoother
for new users.

RezwanAhmed123 committed:
Update Java Doc for TaskEditor.java

The javadoc is based on the old method before refactoring the code.

The doc explains the older format of the method when it dealt with
all of marking, unmarking and deletion. This is confusing for any
new developers of the project.

Let's update the Javadoc such that it reflects the new functions of
the method after refactoring.

This makes the method more understandable for new developers when
they need to modify the method or use the method for other methods
of the program.

RezwanAhmed123 committed:
Refactor handleUserInput method

**Note I was testing a use case to let AI write my commit messages
and this is the result from ChatGPT". I based it on other commit
messages I wrote in the past:

The handleUserInput method was refactored to improve clarity and
error handling when processing user input and executing commands.
The following changes were made:

- Reorganized the dialogContainer's getChildren().addAll() calls
  for better code readability and maintainability.
- Extracted common dialog box addition logic into a single block,
  reducing code duplication.
- Improved error handling by catching MortalReminderException and
  providing more specific feedback when an exception occurs.
- Enhanced the dialog creation by ensuring correct command type
  (known or unknown) is reflected in the mortal reminder dialog.

This update makes the code easier to maintain, reduces redundancy,
and provides clearer messaging for both successful and erroneous
cases.

RezwanAhmed123 committed:
Modify TaskListStorageTest.java

The test is not setting up and cleaning the relevant files before
and after the test.

This leads to issues in passing the test especially if the storage
list file already contained some items in it.

Lets add "@BeforeEach" and "@AfterEach" command for this file to
set up proper testing environments before and after each test case
while also following the DRY (Don't Repeat Yourself) principle.

This allows the testing process to be smoother for further builds
and github continuous integrations.

RezwanAhmed123 committed:
Add test cases for TaskCreator.java

This method is crucial in our program but is not robustly tested.

This may lead to many errors in coding not being correctly caught at
the bud when its dependent methods are modified without care about
the effect on this class.

Lets add more robust test cases for this method to quickly ensure
any regressions are caught timely and modified accordingly.

This should save a lot of time and effort as more modifications are
made for the program.

RezwanAhmed123 committed:
Update AI.md

The AI-assistance documentation has not been updated in a while.

A lot of AI assistance was used for many methods and is difficult to
keep track of.

Lets update the AI.md to quickly document all the ways AI was used
and lessons learnt during the journey.

This will help close learning loop and give credit where its due
to prevent accusations of plagiarism.

RezwanAhmed123 committed:
Update Event.java

There is a regression error that is spotted on Event.java due to
added space before the "/from" inside of the getDescription() method.

This causes one of the test cases to fail which prevents continuous
integration inside github.

Let's fix the regression error inside Event.java.

This will allow the build to be created successfully and CI to be
smooth.

RezwanAhmed123 committed:
Update TimedTask.java

The locale of the DateTimeFormat class was not specified.

When different OS or machines use the program, the program will
follow their specific device locale and this will cause problems for
the storage of the times.

Let's add a specific locale to follow for the program.

This will allow the behaviour of the DateTimeFormat class to be
consistent across many different devices and OS.

RezwanAhmed123 committed:
Modify TaskListCreatorTest.java

There is repetition within the test methods of
TaskListCreatorTest.java.

This reduces the code quality of the testing phase and makes new
methods more difficult to test as the setup will have to be repeated
again.

Let's extract out the setup from each method into a class level
implementation.

This reduces the amount of repetition and allows us to change the
test cases as quickly as possible. This also causes the test to test
the behaviour of each class rather than the execution only.

RezwanAhmed123 committed:
Add test cases for TaskEditorTest.java

This class has not been tested robustly.

This will prevent regressive bugs from being detected early.

Let's add some test cases to automatically test this during the
building or running of the app on gradle.

This will allow regression errors to be caught more swiftly
and modify these bugs before they become a bigger problem down the
line.

RezwanAhmed123 committed:
Refactor code

Code has many magic literals especially error messages.

It looks messy and some of the error messages are not consistent.

Lets extract out the magic literals and keep repetition to a minimum.

This will make the code more readable and consistent.

RezwanAhmed123 committed:
Reformat Code

Refactored magic literals to make the code more readability.

RezwanAhmed123 committed:
Refactor code

Some magic literals were still present in code.

This makes the code inconsistent.

Let's reformat the magic literals and make the code and methods more
consistent.

This also makes our code SLAP harder and is more readable for new
developers down the line.

RezwanAhmed123 committed:
Add personality quips to program

The current feedback messages are too bland.

This may make the app uninteresting to some users.

Let's spice up the feedback messages without confusing the users.

This should help make the app more interesting.

RezwanAhmed123 committed:
Add documentation for ReadMe in docs

The project website is not up.

Lets add a UI picture and set up a simplified version of the project
user guide up onto the project website.

RezwanAhmed123 committed:
Complete Readme.md in docs

The readme is not complete as some features were missing.

Lets complete the readme by adding information about all the missing
features and deploy to the website.

RezwanAhmed123 committed:
Polish Final Version

Everything is ready for release.

Let's polish everything in the ReadMe.md for the final version
(except for the download link which will only be ready once
the release is out).

RezwanAhmed123 committed:
Complete release

The link to the release download is broken.

Let's fix the release download link so users can download the app
and use it themselves.

RezwanAhmed123 committed:
Fix broken section of ReadMe.md

The mark and unmark sections were not rendering correctly in the
website.

Let's fix the code formatting in these sections so the code is
visible in the project website.

RezwanAhmed123 committed:
Minor Edits of ReadMe.md

Some of the sections were not organised properly and the code
sections also did not show properly because of this.

Let's modify the format accordingly so the code can be seen
accurately.

RezwanAhmed123 committed:
Remove dead methods

There is one dead initialisation method in CommandAlternatives class.

This may be confusing for new developers to the project because they
may not understand the reason for the dead method, what purpose it
serves.

Let's remove the dead method from the project.

This will improve code quality and readability for new developers to
the project.

RezwanAhmed123 committed:
Fix find function

Find function returns entire list of tasks instead of correctly
filtered tasks.

This is a bug and will confuse the user and may make them stop using
the app.

Let's fix the bug so the find functionality can be used properly.

This will allow the app to be used seamlessly again.

RezwanAhmed123 committed:
Fix test cases

Some error messages were not reflected correctly and threw errors in
the test cases.

This may confuse new developers who may have changed/fixed other
issues in the code and this throws an error.

Let's fix the issues in the test cases to prevent confusion down the
line.

RezwanAhmed123 committed:
Update Userguide

The link for the user guide is not updated.

Let's update the user guide so the download link is not broken,
allowing the user to download the app properly.
